jax/_src/lax/control_flow/loops.py

```diff
@@ -1029,6 +1031,59 @@ def _scan_pp_rule(eqn, context, settings):
     del printed_params['reverse']
   return core._pp_eqn(eqn.replace(params=printed_params), context, settings)
 
+def _scan_discharge_rule(in_avals, out_avals, *args, jaxpr, num_consts,
+                         num_carry, linear, unroll, reverse, length):
+  jaxpr, consts = jaxpr.jaxpr, jaxpr.consts
+  if consts: raise NotImplementedError
+  consts, carry, xs = split_list(args, [num_consts, num_carry])
+  consts_linear, carry_linear, xs_linear = split_list(
+      linear, [num_consts, num_carry])
+  consts_avals, carry_avals, xs_avals = split_list(in_avals,
+                                                   [num_consts, num_carry])
+  is_ref = [isinstance(a, state.AbstractRef) for a in consts_avals]
+  remaining_const_avals, in_ref_avals = partition_list(is_ref, consts_avals)
+  remaining_consts, in_refs = partition_list(is_ref, consts)
+  remaining_consts_linear, in_refs_linear = partition_list(is_ref, consts_linear)
+  num_refs = sum(is_ref)
+  num_extensive_in = len(in_avals) - num_carry - num_consts
+  num_extensive_out = len(out_avals) - num_carry
+  num_remaining_consts = num_consts - num_refs
+  discharged_jaxpr, discharged_consts = state_discharge.discharge_state(jaxpr, ())
+  if discharged_consts: raise NotImplementedError
+  # The discharged jaxpr will have output refs stashed at the end
+  def wrapped(*refs_and_args):
+    consts, refs, carry, xs = split_list(refs_and_args, [num_remaining_consts,
+                                                         num_refs,
+                                                         num_carry])
+    consts_with_refs = merge_lists(is_ref, consts, refs)
+    outs_and_refs = core.eval_jaxpr(discharged_jaxpr, (), *consts_with_refs,
+                                    *carry, *xs)
+    carry, ys, out_refs = split_list(outs_and_refs, [num_carry,
+                                                     num_extensive_out])
+    assert len(out_refs) == num_refs
+    return [*out_refs, *carry, *ys]
+  new_in_avals = [*remaining_const_avals, *[a.inner_aval for a in in_ref_avals],
+                  *carry_avals,
+                  *[core.mapped_aval(length, 0, a) for a in xs_avals]]
+  new_jaxpr, _, new_consts = pe.trace_to_jaxpr_dynamic(lu.wrap_init(wrapped),
+                                                       new_in_avals)
+  if new_consts: raise NotImplementedError
```

Comment:
I think this may be impossible given that `discharged_jaxpr` has no consts; if so could replace `new_consts` two lines up with `()`.

Reply:
Done.